android/libraries/rib-base/src/androidTest/java/com/badoo/ribs/android/lifecycle/BaseNodesTest.kt

```diff
@@ -1,57 +1,58 @@
-        permanentParts = permanentParts.map { it.toNodeBuilder() },
-        dialogLauncher = dialogLauncher
+    @get:Rule
+    val ribsRule = RibsRule()
+
+    data class When(
+        val permanentParts: List<TestRootRouter.Configuration.Permanent> = emptyList(),
+        val initialConfiguration: TestRootRouter.Configuration.Content = TestRootRouter.Configuration.Content.NoOp,
+        val pushConfiguration1: TestRootRouter.Configuration? = null,
+        val pushConfiguration2: TestRootRouter.Configuration? = null
     )
 
-    protected val router: TestRootRouter
-        get() = rootProvider.rootNode?.getRouter() as TestRootRouter
+    protected fun test(setup: When, expectedState: ExpectedState, testBlock: (TestRootRouter) -> Unit) {
+        val rootProvider = TestRoot.Provider(
+            initialConfiguration = setup.initialConfiguration,
+            permanentParts = setup.permanentParts
+        )
 
-    class NodeState(
-        val attached: Boolean,
-        val viewAttached: Boolean
-    ) {
-        override fun toString() = "attached = $attached, view attached = $viewAttached"
-    }
+        ribsRule.start { rootProvider.create(it.dialogLauncher()) }
 
-    class ExpectedState(
-        val node1: NodeState?,
-        val node2: NodeState?
-    ) {
-        override fun toString() = "*then* node1 should be [$node1] and node2 should be [$node2]"
-    }
+        val router: TestRootRouter = rootProvider.rootNode?.getRouter() as TestRootRouter
+
+        testBlock.invoke(router)
 
-    enum class TestNode {
-        NODE_1,
-        NODE_2
+        rootProvider.makeAssertions(expectedState)
     }
 
-    private fun TestNode.toNodeBuilder(): () -> Node<*> = when (this) {
-        NODE_1 -> ({ rootProvider.childNode1Builder() })
-        NODE_2 -> ({ rootProvider.childNode2Builder() })
+    protected fun TestRootRouter.pushIt(configuration: TestRootRouter.Configuration) {
+        when (configuration) {
+            is TestRootRouter.Configuration.Content -> push(configuration)
+            is TestRootRouter.Configuration.Overlay -> pushOverlay(configuration)
+        }
     }
 
-    protected fun makeAssertions(expected: ExpectedState) {
-        assertThat(rootProvider.childNode1?.isAttached).describedAs("is child node 1 attached").isEqualTo(expected.node1?.attached)
-        assertThat(rootProvider.childNode1?.isViewAttached).describedAs("is child node 1 view attached").isEqualTo(expected.node1?.viewAttached)
-        assertThat(rootProvider.childNode2?.isAttached).describedAs("is child node 2 attached").isEqualTo(expected.node2?.attached)
-        assertThat(rootProvider.childNode2?.isViewAttached).describedAs("is child node 2 view attached").isEqualTo(expected.node2?.viewAttached)
+    private fun TestRoot.Provider.makeAssertions(expected: ExpectedState) {
+        Espresso.onIdle()
+        assertThat(permanentNode1?.isAttached).describedAs("is Permanent 1 Attached").isEqualTo(expected.permanentNode1?.attached)
```

Comment:
in the tests, we use predefined constants like ON_SCREEN/VIEW_DETACHED/etc. In case of failure, these assertions will print something like `is Child 1 Attached is [true] but expected [false]`. Maybe we should make similar assertions here, so the error message will display `Child 1 is [ON_SCREEN] but expected [VIEW_DETACHED]`?

Reply:
Done.